### PR TITLE
Added FIRST Support

### PR DESCRIPTION
**Pull request by wu-ming233**
# Pull Request Checklist

## Reference Issue

Please provide the reference to issue this PR is addressing (# followed by the issue number). If there is no associated issue, write "N/A".

ref: N/A

## Checklist Items

Before submitting your pull request, please review these items:

- [x] Have you followed the [contributing guidelines](CONTRIBUTING.md)?
- [x] Have you verified that there are no existing [Pull Requests](https://github.com/castorini/rank_llm/pulls) for the same update/change?
- [x] Have you updated any relevant documentation or added new tests where needed?

# PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Documentation content changes
- [ ] Other... 

Description: 
- PR corresponding to [FIRST Repro](https://arxiv.org/abs/2411.05508)
- Now supports reranking with models from huggingface, instead of just rank zerphyr and vicuna
- Only loads from huggingface when `vllm_batched` enabled
- Now supports reranking with logits of 1st identifier only
- Now supports using alphabetical identifiers instead of numerical

All the code only support the vllm backend, as the fastchat-related business is outdated. However, the fastchat code were still kept for backwards compatibility, when `vllm_ batched` is not turned on.

Currently, everything is passed in as an optional argument to existing rank_llm functions to enable use_first_logits or use_alphabetial_identifiers. We could think about better abstractions later.

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -2,7 +2,7 @@ tqdm>=4.66.2
 openai>=1.23.6
 tiktoken>=0.6.0
 transformers>=4.40.1
-pyserini>=0.35.0
+pyserini==0.39.0
```
Comment (ronakice) [inline requirements.txt:5]:
```suggestion
pyserini==0.42.0
```

This is an important push and removes extra dependencies, can you remove faiss-cpu and see if it works?

File: src/rank_llm/retrieve_and_rerank.py [hunk 2]
```diff
@@ -37,7 +37,7 @@ def retrieve_and_rerank(
 
     # Get reranking agent
     reranker = Reranker(
-        Reranker.create_agent(model_path.lower(), default_agent, interactive, **kwargs)
+        Reranker.create_agent(model_path, default_agent, interactive, **kwargs)
```
Comment (ronakice) [inline src/rank_llm/retrieve_and_rerank.py:40]: why this change?

File: src/rank_llm/scripts/tmp.ipynb
Comment (ronakice) [inline src/rank_llm/scripts/tmp.ipynb:1]: what's this?

**Review by ronakice — CHANGES_REQUESTED** [review]
Quick fixes

File: requirements.txt (hunk 1 above)
Comment (wu-ming233) [reply requirements.txt]: Pyserini 0.42.0 refactored things that break rank_llm. Currently pushed an ugly solution for rank_llm to work, but would be better to update on pyserini's side as well

File: src/rank_llm/retrieve_and_rerank.py (hunk 2 above)
Comment (wu-ming233) [reply src/rank_llm/retrieve_and_rerank.py:40]: Reddy's First_Model breaks if it's lower-cased

File: src/rank_llm/scripts/tmp.ipynb
Comment (wu-ming233) [reply src/rank_llm/scripts/tmp.ipynb:1]: Sorry about that. Removed
